packages/block-editor/src/hooks/text-align.js

```diff
@@ -0,0 +1,190 @@
+ * Takes into consideration the text aligns supported by a block.
+ * Exported just for testing purposes, not exported outside the module.
+ *
+ * @param {?boolean|string[]} blockTextAlign Text aligns supported by the block.
+ *
+ * @return {string[]} Valid text alignments.
+ */
+export function getValidTextAlignments( blockTextAlign ) {
+	let validTextAlignments;
+	if ( Array.isArray( blockTextAlign ) ) {
+		validTextAlignments = VALID_TEXT_ALIGNMENTS.filter( ( textAlign ) =>
+			blockTextAlign.includes( textAlign )
+		);
+	} else if ( blockTextAlign === true ) {
+		// `true` includes all alignments...
+		validTextAlignments = [ ...VALID_TEXT_ALIGNMENTS ];
+	} else {
+		validTextAlignments = [];
+	}
+
+	return validTextAlignments;
+}
+
+/**
+ * Filters registered block settings, extending attributes to include `textAlign`.
+ *
+ * @param {Object} settings Original block settings.
+ *
+ * @return {Object} Filtered block settings.
+ */
+export function addAttribute( settings ) {
+	// Allow blocks to specify their own attribute definition with default values if needed.
+	if ( 'type' in ( settings.attributes?.textAlign ?? {} ) ) {
+		return settings;
+	}
+	if ( hasBlockSupport( settings, TEXT_ALIGN_SUPPORT_KEY ) ) {
+		// Gracefully handle if settings.attributes is undefined.
+		settings.attributes = {
+			...settings.attributes,
+			textAlign: {
+				type: 'string',
+				// Allow for '' since it is used by the `updateTextAlignment` function
+				// in toolbar controls for special cases with defined default values.
+				enum: [ ...VALID_TEXT_ALIGNMENTS, '' ],
+			},
+		};
+	}
+
+	return settings;
+}
+
+function BlockEditTextAlignmentToolbarControlsPure( {
+	name: blockName,
+	textAlign,
+	setAttributes,
+} ) {
+	const validTextAlignments = getValidTextAlignments(
+		getBlockSupport( blockName, TEXT_ALIGN_SUPPORT_KEY )
+	);
+	const blockEditingMode = useBlockEditingMode();
+	if ( ! validTextAlignments.length || blockEditingMode !== 'default' ) {
+		return null;
+	}
+
+	const textAlignmentControls = TEXT_ALIGNMENT_OPTIONS.filter( ( control ) =>
+		validTextAlignments.includes( control.align )
+	);
+
+	const updateTextAlignment = ( nextTextAlign ) => {
+		if ( ! nextTextAlign ) {
+			const blockType = getBlockType( blockName );
+			const blockDefaultTextAlign =
+				blockType?.attributes?.textAlign?.default;
+			if ( blockDefaultTextAlign ) {
+				nextTextAlign = '';
+			}
+		}
+		setAttributes( { textAlign: nextTextAlign } );
+	};
+
+	return (
+		<BlockControls group="block">
+			<AlignmentControl
+				value={ textAlign }
+				onChange={ updateTextAlignment }
+				alignmentControls={ textAlignmentControls }
+			/>
+		</BlockControls>
+	);
+}
+
+export default {
+	edit: BlockEditTextAlignmentToolbarControlsPure,
+	useBlockProps,
+	addSaveProps: addAssignedTextAlign,
+	attributeKeys: [ 'textAlign' ],
+	hasSupport( name ) {
+		return hasBlockSupport( name, TEXT_ALIGN_SUPPORT_KEY, false );
+	},
+};
+
+function useBlockProps( { name, textAlign } ) {
```

Comment:
For consistency with other block supports and supporting more complex third-party blocks, shouldn't this handle skipping serialization?

e.g. [font family support](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/font-family.js?plain=1#L56)